### DISCUSSION TITLE
What are the roles of the RPC methods, defined in proto file?

### DISCUSSION THREAD

**Discussion by Prateeknandle** [evilsocket/opensnitch#1164] [Development]
I found there are five RPC calls defined in proto, but there is no mention of which RPC provides what kind of service in the documentation.

I was building a service which wants to connect with the daemon (ui is not present), and it seems out of 5 RPC calls only `Notifications` is the one which streams data, but I do not find daemon is sending any data to it. I need an alerts/events stream where any deny action or allowed action takes place, defined by the rules, the service will get the alerts/events. 

Is this even possible with current codebase?

**Comment by gustavo-iniguez-goya** [comment]
Hi @Prateeknandle !

The Notifications channel is mainly used to send "Actions" from the server (UI) to the daemon, and receive the result of the action:
Apply/Change/delete application or firewall rules, change configuration, etc.

https://github.com/evilsocket/opensnitch/blob/1984fb995410044b89aad4c4396479ee13c62094/daemon/ui/notifications.go#L270

daemon -> (daemon/ui/client) Subscribe() -> server (UI).

UI -> Apply rule -> Notifications() <<(grpc channel)>> -> daemon -> handleNotification() -> protocol.Action_CHANGE_RULE -> result Ok/Error

> I need an alerts/events stream where any deny action or allowed action takes place, defined by the rules, the service will get the alerts/events.

Those events are sent via the Ping() gRPC method. The daemon collects statistics about the monitored events (connections blocked/allowed), and send them to the server (UI) every second.

new connection -> (daemon/statistics/)  OnConnectionEvent -> add to circular buffer
https://github.com/evilsocket/opensnitch/blob/1984fb995410044b89aad4c4396479ee13c62094/daemon/statistics/stats.go#L119
https://github.com/evilsocket/opensnitch/blob/1984fb995410044b89aad4c4396479ee13c62094/daemon/statistics/stats.go#L192

at the same time, the collected stats are sent to the server (UI): (daemon/ui/client.go) Client.ping() -> stats.Serialize()

https://github.com/evilsocket/opensnitch/blob/1984fb995410044b89aad4c4396479ee13c62094/daemon/ui/client.go#L300
https://github.com/evilsocket/opensnitch/blob/1984fb995410044b89aad4c4396479ee13c62094/daemon/statistics/stats.go#L255

and every time the collected stats are sent, they're deleted from the daemon.

**Reply by gustavo-iniguez-goya** [reply]
Alternatively, you can send the network events to a remote server, by using Loggers:

https://github.com/evilsocket/opensnitch/wiki/SIEM-integration

**Comment by Prateeknandle** [comment]
Hey thanks @gustavo-iniguez-goya for detailed explanation, much appreciate your help.
